include/zephyr/drivers/espi_emul.h

```diff
@@ -144,42 +126,61 @@ int espi_emul_register(const struct device *dev, const char *name,
 		       struct espi_emul *emul);
 
 /**
- * Sets the eSPI virtual wire on the host side, which will
- * trigger a proper event(and callbacks) on the emulated eSPI controller
- *
- * @param espi_dev eSPI emulation controller device
- * @param vw The signal to be set.
- * @param level The level of the signal to be set.
+ * Perform the I/O read through the eSPI bus
  *
- * @retval 0 If successful.
- * @retval -EIO General input / output error.
+ * @param dev Device that will handle the operation
+ * @param length number of bytes to read (1/2/4 supported)
+ * @param addr I/O address to read
+ * @param reg Pointer to value which will store the data
+ * @return 0 if IO was successful, -EINVAL if address is not supported
  */
-int emul_espi_host_send_vw(const struct device *espi_dev,
-			   enum espi_vwire_signal vw, uint8_t level);
+int espi_emul_io_read(const struct device *dev,
+	uint8_t length, uint16_t addr, uint32_t *reg);
 
 /**
- * Perform port80 write on the emulated host side, which will
- * trigger a proper event(and callbacks) on the emulated eSPI controller
+ * Perform the I/O write through the eSPI bus
  *
- * @param espi_dev eSPI emulation controller device
- * @param data The date to be written.
+ * @param dev Device that will handle the operation
+ * @param length number of bytes to write (1/2/4 supported)
+ * @param addr I/O address to write
+ * @param reg Data to write
+ * @return 0 if IO was successful, -EINVAL if address is not supported
+ */
+int espi_emul_io_write(const struct device *dev,
+	uint8_t length, uint16_t addr, uint32_t reg);
+
+/**
+ * Set the internal value of virtual-wire on the host side.
  *
- * @retval 0 If successful.
- * @retval -EIO General input / output error.
+ * @param dev Device that will handle the operation
+ * @param vw virtual wire which value will be changed
+ * @param level logic level on virtual wire
+ * @return 0 if IO was successful, -EIO if there's no emulator, -EPERM if
+ *           the direction of vwire is invalid
  */
-int emul_espi_host_port80_write(const struct device *espi_dev, uint32_t data);
+int espi_emul_internal_set_vwire(const struct device *dev,
```

Comment:
What does "internal" mean here? Should this be part of the API for this module?

Reply:
It means that internal in the emulation sense. Let's think about the host, it emulates that it's internal value of virtual wire has changed, and it has to send this information through the eSPI bus to "us" (EC).

Reply:
So you are distinguishing between the current value of the virtual wire as reported to the EC (`send_vwire`?) and the value that the emulated host has attempted to set (`internal_set_vwire`)?

Reply:
Not exactly. The send_vwire function sends the virtual wire value from the EC to the other side, e.g. host. It's used to inform the host about change in wire connected to the EC. The receive_vwire is used by EC to read the value of wire connected to the host. So this internal_set_vwire is like a send_vwire function but from the host side to the EC. If the AP wants to go to sleep mode, it sends the vwire slp signals through the eSPI to the EC. This internal function emulates this.

Reply:
Would `host_set_vwire` make sense as a name? To me at least, that would be clearer.

Reply:
It would make much sense in this case. The question is, if our "secondary" (host) side will always be a host? Maybe it is a good idea to change that, even if there will be other emulators in the future,

Reply:
In real life, do the EC and on-chip keyboard controller communicate over the eSPI bus or using some other channel inside the EC chip? If there was an off-chip keyboard controller, would it talk to the EC at all?

If the EC and keyboard controller wouldn't actually communicate over the bus, then maybe there should be two layers of API or an otherwise slightly different structure here. (I have been kind of wondering why the keyboard controller emulation was part of the eSPI emulation module, so thank you for elaborating.)

Reply:
IIUC, eSPI is specifically intended to be used in the same context as LPC, which is intended for CPUs to communicate with peripherals on motherboards. So I think unless Zephyr is running on the CPU, the "other" side of this emulator will always be the CPU. To the extent that we use the API to communicate with the keyboard controller, it's outside the bounds of eSPI as a technology, and we can probably expect that we won't need to make similar accommodations for anything else.

I would prefer `host_set_vwire` as a name for this function and API function pointer. "cpu" or "ap" would also be reasonable choices. I don't think "secondary" or "internal" convey clearly what this function does. Same idea for the comments that use "secondary."